pkg/providers/openai/chat.go

```diff
@@ -1,53 +1,262 @@
+	Object  string        `json:"object,omitempty"`
+	Usage   struct {
+		CompletionTokens float64 `json:"completion_tokens,omitempty"`
+		PromptTokens     float64 `json:"prompt_tokens,omitempty"`
+		TotalTokens      float64 `json:"total_tokens,omitempty"`
+	} `json:"usage,omitempty"`
+}
+
+// Chat sends a chat request to the specified OpenAI model.
+//
+// Parameters:
+// - payload: The user payload for the chat request.
+// Returns:
+// - *ChatResponse: a pointer to a ChatResponse
+// - error: An error if the request failed.
+func (c *ProviderClient) Chat() (*ChatResponse, error) {
+	// Create a new chat request
+
+	slog.Info("creating chat request")
+
+	chatRequest := c.CreateChatRequest(c.Payload)
+
+	slog.Info("chat request created")
+
+	// Send the chat request
+
+	slog.Info("sending chat request")
+
+	resp, err := c.CreateChatResponse(context.Background(), chatRequest)
+
+	return resp, err
+}
+
+func (c *ProviderClient) CreateChatRequest(message []byte) *ChatRequest {
+	var requestBody providers.RequestBody
+	err := json.Unmarshal(message, &requestBody)
+	if err != nil {
+		slog.Error("Error:", err)
+		return nil
+	}
+
+	slog.Info("creating chatRequest from payload")
+
+	var messages []*ChatMessage
+	for _, msg := range requestBody.Message {
+		chatMsg := &ChatMessage{
+			Role:    msg.Role,
+			Content: msg.Content,
+		}
+		if msg.Role == "user" {
+			chatMsg.Content += " " + strings.Join(requestBody.MessageHistory, " ")
+		}
+		messages = append(messages, chatMsg)
+	}
+
+	// iterate through self.Provider.DefaultParams and add them to the request otherwise leave the default value
+
+	chatRequest := &ChatRequest{
+		Model:            c.setModel(),
+		Messages:         messages,
+		Temperature:      0.8,
 		TopP:             1,
+		MaxTokens:        100,
 		N:                1,
+		StopWords:        []string{},
 		Stream:           false,
-		Stop:             nil,
-		PresencePenalty:  0,
 		FrequencyPenalty: 0,
+		PresencePenalty:  0,
 		LogitBias:        nil,
 		User:             nil,
 		Seed:             nil,
-		Tools:            nil,
+		Tools:            []string{},
 		ToolChoice:       nil,
 		ResponseFormat:   nil,
 	}
+
+	// Use reflection to dynamically assign default parameter values
+	defaultParams := c.Provider.DefaultParams
+
+	chatRequestValue := reflect.ValueOf(chatRequest).Elem()
+	chatRequestType := chatRequestValue.Type()
+
+	for i := 0; i < chatRequestValue.NumField(); i++ {
+		jsonTags := strings.Split(chatRequestType.Field(i).Tag.Get("json"), ",")
+		jsonTag := jsonTags[0]
+		if value, ok := defaultParams[jsonTag]; ok {
+			fieldValue := chatRequestValue.Field(i)
+			fieldValue.Set(reflect.ValueOf(value))
+		}
+	}
+
+	return chatRequest
+}
+
+// CreateChatResponse creates chat Response.
+func (c *ProviderClient) CreateChatResponse(ctx context.Context, r *ChatRequest) (*ChatResponse, error) {
+	_ = ctx // keep this for future use
+
+	resp, err := c.createChatHTTP(r) // netpoll/hertz does not yet support tls
```

Comment:
I'm wondering how hertz is connected to our provider clients? Hertz powers Glide API. This means it's involved when Glide's users make request to Glide to serve those request which is to dispatch the corresponding handler by route & to validate the incoming schemas there. The validation would produce a unified request schema object that we operated on in the request of the gateway.

This point is far too deeply inside of the gateway logic to interact with hertz.

Reply:
Yep, you're correct. I wasn't totally sure but Hertz is not involved.